pkg/virt-controller/watch/migration.go

```diff
@@ -522,6 +522,19 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
 			// migration was accepted into the system, now see if we
 			// should create the target pod
 			if vmi.IsRunning() {
+				vmiCopy := vmi.DeepCopy()
+				vmiCopy.Status.MigrationState = &virtv1.VirtualMachineInstanceMigrationState{
+					MigrationUID: migration.UID,
+					SourceNode:   vmi.Status.NodeName,
+				}
+				if !reflect.DeepEqual(vmi.Status, vmiCopy.Status) ||
+					!reflect.DeepEqual(vmi.Labels, vmiCopy.Labels) {
+					_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Update(vmiCopy)
```

Comment:
I think we should use a Patch instead of an Update here perhaps

Reply:
Yes. There are *so* many places in the code where we should use Patch instead of Update... This is copy-pasted from the following case... Yes I should fix them all, but the original intent was to keep this PR as short and backport-able as possible.

What you are suggesting in your main comment still allows the creation of many migrations for 1 VMI.
IMO this is inherently wrong, the admitter should really only let the first one in and reject everything else.
Even if this PR lives in migration.go, the intent is really to alter the behavior of the admitter.

That being said, going back to "short and backport-able", your suggested approach makes a lot of sense, and should be enough to fix the issue. I can do that in a separate PR, if that works for you, since this one still needs to exist to solve the bigger issue.

Reply:
I am struggling with this...
The best I could do is serialize all the migrations created, which doesn't really buy us anything. (https://github.com/jean-edouard/kubevirt/commit/b1eb3e7f329f78098babcf9db70e425a5bbe92e5)
I may give this another try in the future, unless someone else does.

Reply:
>  I can do that in a separate PR, if that works for you, since this one still needs to exist to solve the bigger issue.

how does this PR address the creation of many migrations for 1 VM? that's an issue we'll always have. we can make a best effort attempt to solve it in the webhook, and then our controller needs to know how to handle the edge case where multiple concurrent migrations exist by only allowing one to succeed (the rest are moved to failed)

> The best I could do is serialize all the migrations created, which doesn't really buy us anything.

@jean-edouard your logic is pretty close, here's what i was thinking. This would all take place in the
```
c.migrationStartLock.Lock()
defer c.migrationStartLock.Unlock()
```
critical section that already exists for virtv1.MigrationPending


```
func (c *MigrationController) canCreateTarget(migration *virtv1.VirtualMachineInstanceMigration, vmi *virtv1.VirtualMachineInstance) (bool, error) {
	
        // get all known migrations in the VMI's namespace
	objs, err := c.migrationInformer.GetIndexer().ByIndex(cache.NamespaceIndex, vmi.Namespace)
	if err != nil {
		return false, err
	}


        // find the ones that match the VMI, if a migration is found that isn't finalized, see if it has any active pods, if so bail.
	for _, obj := range objs {
		curMigration := obj.(*virtv1.VirtualMachineInstanceMigration)
		if curMigration.Spec.VMIName != vmi.Name {
			continue
		} else if string(curMigration.UID) == string(migration.UID) {
                        continue
               } else if curMigration.IsFinal() {
                       continue
               }

               // curMigration is inflight and does not match the migration the sync loop is working on. 
               targetPods, err = c.listMatchingTargetPods(migration, vmi)
               for _, targetPods := range targetPods {
                       if targetPod.Phase != k9sv1.PodSucceeded && targetPod.Phase != k8sv1.PodFailed {
                                // another migration exists with an active target pod. 
                                return false, nil
                       }
               }
	}

        return true, nil
```

Reply:
That's a much better version of the commit above, but still seems like a weird way to go about this... More below.

Reply:
I think there was some confusion here, `canCreateTarget` returning false will not cancel the migration, just delay it.
Sounds like if we manage to make it cancel it instead we'll be in good shape!
Also in the code above, shouldn't `c.listMatchingTargetPods(migration, vmi)` read `c.listMatchingTargetPods(curMigration, vmi)`?
Thanks!

Reply:
yeah, lets just do it your way with using the VMI's status as the gating mechanism.  We'd basically be committing to processing the migration if we can set the vmi.Status.MigrationState... If we can't set that on the VMI because another migration is in-flight, then the migration being processed should fail early due to another migration in flight.